Add optional Consent model (Close #126)

agnessnowplow opened the pull request:
## Description & motivation
Adding the custom module for processing consent tracking events. The reason why it is still in a draft PR is because discussion is still going on about adding a potential new boolean attribute called `gdpr_applies` to the schema and also because there has been a recent schema renaming (`action` will be renamed to `event_type` so that I do not have to rename it in the model right away as it is now) but once finalised and I can actually see events from the final schema I will change it. Testing currently can only be done in BigQuery, but Peter and Jack are working on sending some data through snowflake atm so will keep you updated. Some info for ease of access:

```
vars:
  snowplow_web:
    snowplow__start_date: '2022-10-28'
    snowplow__backfill_limit_days: 10
    snowplow__atomic_schema: 'rt_pipeline_dev1'

```
An FYI, I left the module disabled by default like so:

```
models:
  snowplow_web:
    optional_modules:
      consent:
        enabled: false
        bigquery:
          enabled: "{{ target.type == 'bigquery' | as_bool() }}"
        snowflake:
          enabled: "{{ target.type == 'snowflake' | as_bool() }}"
```
Disabling only on consent level should be enough after all because enabling from the user's project overwrites the lower level enabling so they would have to add all these code anyway. Not yet removed though, this just came to my mind.

## Checklist
- [ ] I have verified that these changes work locally

snowplowcla commented:
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

rlh1994 commented on models/optional_modules/consent/snowplow_web_consent_totals.sql:
```diff
@@ -0,0 +1,50 @@
+          and {{ dbt_utils.dateadd('year', '1', 'last_consent_event_tstamp') }} > current_date then 1 end) as expires_in_six_months
+
+  from {{ ref('snowplow_web_consent_users') }}
+
+  where is_official_version
+
```
❓ / 💡 Could there be some value in this table containing information about the previous versions as well? So the first query becomes a distinct and then the join is on the version? It might be useful to see how many users they still have on previous versions - or is that not something that makes sense in the context of cmps?

agnessnowplow replied:
As far as I understand whatever is the 'official', latest consent version that's all that matters from a compliance point of view. I can imagine from specific analytical purposes it could be useful to also see the previous version though so there's not much to lose by adding it there tbh.

rlh1994 replied:
Makes sense, think it would be good to add if you can, after years of someone suddenly wanting to see some analysis without the filter _they_ asked for I'm just a little cautious around throwing away data that could be useful 😃

rlh1994 commented on models/optional_modules/consent/consent.yml:
```diff
@@ -0,0 +1,134 @@
+        description: The privacy policy url in connection with the last consent event
+      - name: last_domains_applied
+        description: The domains for which the last consent event applies
+      - name: is_official_version
+        description: A boolean to filter whether the last consent or cmp visible event is sent after the latest privacy policy version goes live
```
❓ Is this technically true? It's based off the users table and users first filters out cmp_visible events?

agnessnowplow replied:
Not, unless we instruct them to send a test event, where we do not care about losing the related cmp_visible event. I know this is very ugly. Still thinking with tracker team but if educated they should be good to go.

emielver commented on macros/unnest.sql:
```diff
@@ -0,0 +1,32 @@
+{#
+ Takes care of harmonising cross-db unnesting.
+ #}
+
+{%- macro unnest(id_column, unnest_column, field_alias, source_table) -%}
+    {{ return(adapter.dispatch('unnest', 'snowplow_web')(id_column, unnest_column, field_alias, source_table)) }}
+{%- endmacro -%}
+
+{% macro default__unnest(id_column, unnest_column, field_alias, source_table) %}
+    select {{ id_column }}, explode({{ unnest_column }}) as {{ field_alias }}
+    from {{ source_table }}
+{% endmacro %}
+
+{% macro bigquery__unnest(id_column, unnest_column, field_alias, source_table) %}
+    select {{ id_column }}, r as {{ field_alias }}
+    from {{ source_table }} t, unnest(t.{{ unnest_column }}) r
+{% endmacro %}
+
+{% macro snowflake__unnest(id_column, unnest_column, field_alias, source_table) %}
+    select t.{{ id_column }}, r.value as {{ field_alias }}
+    from {{ source_table }} t, table(flatten(t.{{ unnest_column }})) r
+{% endmacro %}
+
+{% macro postgres__unnest(id_column, unnest_column, field_alias, source_table) %}
+    select {{ id_column }}, cast(trim(unnest({{ unnest_column }})) as {{ dbt_utils.type_int() }}) as {{ field_alias }}
+    from {{ source_table }}
+{% endmacro %}
+
+{% macro redshift__unnest(id_column, unnest_column, field_alias, source_table) %}
+    select {{ id_column }}, {{ field_alias }}
+    from {{ source_table }} p, p.{{ unnest_column }} as {{ field_alias }}
+{% endmacro %}
```
Yes definitely, I've needed pretty much the same exact thing for ecom 😅

agnessnowplow replied:
same for the media player where it already exists, one of the topics I wanted to ask just forgot to put in the PR

agnessnowplow commented on models/optional_modules/consent/bigquery/snowplow_web_consent_log.sql:
```diff
@@ -0,0 +1,92 @@
+    partition_by = snowplow_utils.get_partition_by(bigquery_partition_by = {
+      "field": "derived_tstamp",
+      "data_type": "timestamp"
+    }),
```
Yes, the query tag is only needed for Snowflake as it was asked to do so so that the volume of Snowplow run queries are identified.

rlh1994 replied:
Ah sorry I meant the ones like `tags=["this_run"]`

agnessnowplow replied:
Ah, I see. Not really as it is not a this run table (it is based on a this run table and is already an incremental derived model), all of them are derived which I tagged within the project.yml level so we should be good.

rlh1994 replied:
That's just the schema isn't it? I'm comparing to something like this https://github.com/snowplow/dbt-snowplow-web/blob/c0545dd047842f476f8a6b7d4e9a269c95ed6a12/models/users/snowplow_web_users.sql#L14  but that might not be the same...

agnessnowplow replied:
It is indeed! I will add "derived" to "snowplow_web_incremental" that is already there so it will take effect for all models

agnessnowplow commented on models/optional_modules/consent/bigquery/snowplow_web_consent_log.sql:
```diff
@@ -0,0 +1,92 @@
+
+  from prep p
+
+  where {{ snowplow_utils.is_run_with_new_events('snowplow_web') }} --returns false if run doesn't contain new events.
```
Sure, makes sense to shift it up.

agnessnowplow commented on models/optional_modules/consent/consent.yml:
```diff
@@ -0,0 +1,134 @@
+
+models:
+  - name: snowplow_web_consent_log
+    description: Incremental table showing the audit trail of consent and cmp events
```
Not sure I understand this bit. Do you mean the scratch schema due to the source? `from {{ ref("snowplow_web_base_events_this_run") }} as e` It is coming from the base module that they would not be able to disable so I thought we should be good (?).

rlh1994 replied:
Sorry, I meant the actual tracking schema(s) for the consent stuff, on iglu?

agnessnowplow replied:
ah, ok. I guess I'll have to add it to src_base.yml instead? Not sure where people would expect this.

agnessnowplow commented on models/optional_modules/consent/consent.yml:
```diff
@@ -0,0 +1,134 @@
+      - name: consent_version
+        description: The version of the privacy policy.
+      - name: consent_scopes
+        description: The scopes that the visitor has consented to. The minimum length can be 1 if we suppose that you always have a part of necessary cookies or storage that needs to take be used for the website to work properly
```
I took it directly from the schema definitions, will pass it to the guys, too.

agnessnowplow commented on models/optional_modules/consent/consent.yml:
```diff
@@ -0,0 +1,134 @@
+            - unique
+            - not_null
+        description: '{{ doc("col_domain_userid") }}'
+      - name: user_id
+        description: The optional userid of a visitor
```
I did and it did not find it :D

agnessnowplow replied:
but I can see it being there now so it must have been a typo earlier, will fix that

agnessnowplow commented on models/optional_modules/consent/snowflake/snowplow_web_consent_log.sql:
```diff
@@ -0,0 +1,55 @@
+    upsert_date_key='derived_tstamp',
+    cluster_by=snowplow_web.web_cluster_by_fields_consent(),
+    sql_header=snowplow_utils.set_query_tag(var('snowplow__query_tag', 'snowplow_dbt'))
+  )
```
same reason here too.

agnessnowplow commented on dbt_project.yml:
```diff
@@ -91,6 +93,15 @@ models:
+        enabled: false
+        +schema: "derived"
+        +tags: "snowplow_web_incremental"
+        bigquery:
+          enabled: "{{ target.type == 'bigquery' | as_bool() }}"
+        snowflake:
+          enabled: "{{ target.type == 'snowflake' | as_bool() }}"
```
oh yes you are right, not safe to leave it like that, will definitely remove it (was already a question if we actually need it there as I commented in the intro above)

agnessnowplow commented on integration_tests/dbt_project.yml:
```diff
@@ -35,6 +35,10 @@ models:
+  snowplow_web:
+    optional_modules:
+      consent:
+        enabled: false
```
yes, it is being changed as we speak from the tracking side and would not delay the accelerator due to this but will add an issue to fix it for a new version

agnessnowplow commented on models/optional_modules/consent/snowplow_web_cmp_stats.sql:
```diff
@@ -0,0 +1,85 @@
+    derived_tstamp,
+    event_name,
+    event_type,
+    lag(event_name, 1) over(partition by domain_userid order by derived_tstamp) as lag_type,
+    lag(event_id, 1) over(partition by domain_userid order by derived_tstamp) as lag_id
```
fine by me

agnessnowplow commented on models/optional_modules/consent/snowplow_web_cmp_stats.sql:
```diff
@@ -0,0 +1,85 @@
+    derived_tstamp,
+    event_name,
+    case when event_name = 'cmp_visible' then event_id
+      when lag_type = 'cmp_visible' then lag_id
+      else lag_id end as cmp_id
```
I know the whole code may look like one long messy nonsense upon initial look. I will put some more comments to make it easier to understand what's going on and if you have any ideas to refactor it, please share. The original overall idea was to associate the event_id of the `cmp_visible` event with subsequent consent events until there is another `cmp_visible` event. Once we have this grouping it is possible to find the first consent event after each `cmp_visible` event. Does it make sense? And you are right, it does not work as is currently. I think the logic could become NULL for the else part as we only really need the first event after each `cmp_visible` event anway. Will test it with that and update it like that if it works.

rlh1994 replied:
Ah gotcha, I've had to do this before and remember it being a pain... let me see if I can remember which way I did it and then see how database agnostic it is...

rlh1994 replied:
Would
```sql
LAST_VALUE(CASE WHEN event_name = 'cmp_visible' THEN null ELSE event_id END IGNORE NULLS) OVER (PARTITION BY domain_userid order by derived_tstamp ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW as cmp_id
```
work and remove the need for the lag entirely? It should (I think...) find that last non-null event id (so only events of cmp_visible) _before_ the current record. Only thing I'm always a little iffy on with `ROWS BETWEEN` is iF it includes the current row or not i.e. will a `cmp_visible` event have it's own event_id or the previous one - you'll need to check that manually is probably easiest.

agnessnowplow commented on models/optional_modules/consent/snowplow_web_cmp_stats.sql:
```diff
@@ -0,0 +1,85 @@
+
+  on o.cmp_id = l.event_id
+
+  where l.event_name = 'cmp_visible' and (o.row_num = 2 or o.derived_tstamp = l.derived_tstamp)
```
We are gathering all the cmp_visible events and the next consent_event after that (if there is any). Above you will find the overall concept in my previous comment.

rlh1994 replied:
Makes sense, I hadn't quite registered the join wasn't on event_id on both sides. Just for my own curiosity - what happens if the user gets the cmp_visible then just leaves the site, do they just not have a first consent event? So their records just don't get in the output table?

agnessnowplow replied:
They should still receive a 'pending' consent event. Not sure if it will be enforced or rather just illustrated as the recommended tracking action but I will follow this logic for modelling regardless.

agnessnowplow commented on models/optional_modules/consent/snowplow_web_consent_users.sql:
```diff
@@ -0,0 +1,101 @@
+
+)
+
+, official_consent as (
+
+  select
+
+    consent_version as official_consent_version,
+    derived_tstamp as official_start_date
+
+  from {{ ref('snowplow_web_consent_log') }}
+
+  where event_type = 'allow_all'
+
+  order by derived_tstamp desc
+
+  limit 1
+
+)
```
I raised the exact same concern with the tracker team earlier and their view was that they cmp_visible events are independent from the consent events. Then we agreed to put it in the manual that they have to send a test event right away for the model to work properly without losing some initial data. I have just suggested that perhaps we can configure the trackers to send a test event automatically somehow upon changing a version but not sure if it is possible.

rlh1994 replied:
Hm... could we maybe filter on where those columns't aren't null instead? That way we at least reduce the risk that they need an `allow_all` but it still feels a little wrong...

Also if we stick with this we probably want to ensure that the `_consent_log` table is partitioned/clustered/whatever by dervied_tstamp as we are going to be ordering by it for the whole table every time this is run.

agnessnowplow replied:
it won't work as I do not just need the official version but also the official list of scopes, which will only appear in the allow_all events. The tables are already clustered/partitioned by that I think.

agnessnowplow replied:
By tableS I mean the log in Snowflake or BigQuery

agnessnowplow commented on models/optional_modules/consent/snowplow_web_consent_users.sql:
```diff
@@ -0,0 +1,101 @@
+  left join latest_consents l
+  on c.domain_userid = l.domain_userid
+
+  left join all_consents a
+  on c.domain_userid = a.domain_userid
```
true, will add the missing field, good spot!

agnessnowplow commented on models/optional_modules/consent/snowplow_web_consent_users.sql:
```diff
@@ -0,0 +1,101 @@
+{{
+  config(
+    materialized='table',
+    sql_header=snowplow_utils.set_query_tag(var('snowplow__query_tag', 'snowplow_dbt'))
+  )
+}}
+
```
In theory the log table should not become that massive as it should be a handful per user for a long period of time. However, there ARE unknowns on how this would actually work in real life so I could make it a native dbt incremental table just to be safe without a huge amount of effort on this table so that's fine.

agnessnowplow commented on models/optional_modules/consent/snowplow_web_consent_users.sql:
```diff
@@ -0,0 +1,101 @@
+
+  select
+
+    consent_version as official_consent_version,
+    derived_tstamp as official_start_date
```
No strong opinion on this one, can rename it to `latest`

agnessnowplow commented on models/optional_modules/consent/consent.yml:
```diff
@@ -0,0 +1,134 @@
+      - name: event_name
+        description: '{{ doc("col_event_name") }}'
+      - name: event_type
+        description: The type of consent event. E.g allow_all
```
Just fixed this with the team. Will add the list as a requirement.

agnessnowplow commented:
There is quite an extensive list of changes, tried to split them up to different commits hoping it helps. If I left something out or you need a specific explanation for the bigger refactors, just let me know.

rlh1994 commented on models/optional_modules/consent/snowplow_web_consent_users.sql:
```diff
@@ -0,0 +1,148 @@
+
+  from {{ ref('snowplow_web_consent_log') }}
+
+  where
+  -- enough time has passed since the event happened (derived_tstamp) to be able to process it as a whole (please bear in mind the late arriving data)
+  derived_tstamp < cast({{ dbt_utils.dateadd('hour', var("snowplow__consent_late_arriving_window", 10), dbt_utils.current_timestamp_in_utc() ) }} as {{ dbt_utils.type_timestamp() }})
+
+  {% if is_incremental() %} -- and it has not been processed yet
```
❓ I'm not sure I quite understand this - are we saying we don't want to process consent events that were sent within the last 10 hours? Why would we want to do that? Also from a naming point of view, compared to `snowplow__days_late_allowed` which increases the window we look at this decreases it (I think?), so maybe a name like `snowplow__consent_delay_analysis_window_hours` or something might be clearer about the difference?

Also is there a reason you aren't using this in the `models/optional_modules/consent/snowplow_web_consent_cmp_stats.sql` model?

agnessnowplow replied:
So the idea behind this was that we wait until we are fine to exclude late arriving data and only process things once, otherwise we won't be able to sum up the aggregates. The same concept is in the media player package and my heart just stopped as I believe the logic is wrong, that's why you were confused, too. It should be -10, I believe, and it is currently wrongly calculated within the MP package (aka the Pageview is processed even if it is not yet ready) but please tell me I'm wrong here..

rlh1994 replied:
I don't think you are wrong...

rlh1994 replied:
Worth having someone else triple check!

agnessnowplow replied:
I could actually see it taking effect with the latest snowflake data coming in. I got the data around mid-day and was surprised to see data being modelled as 10 hours have not passed since we got the data. After I changed it to -10 it was not processed and with -1 I was able to process them.. :(

agnessnowplow replied:
On that note, I think we should probably change the logic instead of making this value a negative (aka adding the hours to the derived_tstamp) so that the variable naming make sense to users.

rlh1994 replied:
Yep agree with that, I still think something like `delay` or `buffer` might make it more obvious and distinct to the variable we already have with late, but I'm not too fussed and would rather just align with the media player

agnessnowplow replied:
Renamed it to buffer, the media player has a specific convention due to it processing full pageview events so we are free to rename it as we wish tbf. Just noticed you had another question on why I did not make the cmp_stats incremental. It's cause of the weird lag/lead window shifts, it has to be processed as a table for it to reliably calculate these kpis, I think.

rlh1994 commented on models/optional_modules/consent/snowplow_web_consent_versions.sql:
```diff
@@ -0,0 +1,91 @@
+
+  where event_name <> 'cmp_visible' and event_type = 'allow_all'
+
+  -- enough time has passed since the event happened (derived_tstamp) to be able to process it as a whole (please bear in mind the late arriving data)
+  and derived_tstamp < cast({{ dbt_utils.dateadd('hour', var("snowplow__consent_late_arriving_window", 10), dbt_utils.current_timestamp_in_utc() ) }} as {{ dbt_utils.type_timestamp() }})
+
```
❓ Does this mean that if they make their `allow_all` event we ask them to at the start they won't see it in this table for 10 hours?

agnessnowplow replied:
Unfortunately yes. Another' ugly' bit to work around tracking. It will be emphasised in the tracking implementation manual, probably with this in mind to have this a day before the go live or sth.

rlh1994 reviewed (commented):
Looks awesome and yeah splitting up those commits made it a lot easier thank you! Just a question around the new variable and one suggestion to reduce the CTEs in the users table, otherwise all looking good!

agnessnowplow commented:
Update: we have some events sent to snowflake: SELECT *  FROM SNOWPLOW_DEV1.ATOMIC.events
WHERE collector_tstamp > '2022-11-10' AND app_id = 'consent_accel' Although the snake case field needs to be changed once the final schema version is finalised and new data is sent through. I have found some issues while testing the new snowflake data, most are self explanatory and one I will explain reacting to your earlier comment on the incremental updates.

agnessnowplow commented on models/optional_modules/consent/snowplow_web_consent_users.sql:
```diff
@@ -0,0 +1,148 @@
+
+left join event_aggs a
+on b.domain_userid = a.domain_userid
+
```
Thanks for the suggestion, I refactored it now similarly to what you suggested it.

agnessnowplow commented:
I have received some more test data so I could debug more and identified a couple of issues along the way which I fixed in the latest commits.

Previously the latest version logic was not working properly and I decided to refactor that model to be easier to debug and understand but this way if the users do not keep versioning proper (aka they change a consent scope but keep the same version, which they should not do but it happened with the test data I received) then it will allow for duplicates. Hence I added an extra test for that instead of trying and fixing it in the model itself. I also caught some casting issues for bigquery and missing definitions in the yml. I hope all makes sense. 

I have one last pending point and that is around naming the event_type parameter still. As soon as I get the confirmation I will fix it for both BigQuery and Snowflake, if needed. Otherwise I consider it finished unless I catch something else still. Sorry for the multiple rounds.

rlh1994 commented on models/optional_modules/consent/snowflake/snowplow_web_consent_log.sql:
```diff
@@ -43,13 +45,15 @@ select
   array_to_string(p.domains_applied, ', ') as domains_applied,
-  p.cmp_load_time
+  coalesce(p.gdpr_applies, false) as gdpr_applies,
+  round(p.cmp_load_time, 1) as cmp_load_time
```
⛏️ Is this only recorded to 1 decimal place? Otherwise I would avoid rounding where possible outside of a reporting context.

agnessnowplow replied:
For general use was it does not necessarily make sense to interpret this in a lower value so tracking could have been enough with a less precise value IMHO and the original values come with a long list of decimals hence the conversion. Can remove this for any scientific analysis options though, as rounding was already place downstream.

rlh1994 replied:
If we have a good idea of the true precision of the tracking and could round to that - that would be best. If we have no idea then suggest we leave it as raw.

agnessnowplow replied:
best to remove it at this stage

rlh1994 commented on models/optional_modules/consent/snowplow_web_consent_users.sql:
```diff
@@ -12,48 +12,42 @@ with base as (
-
   {% if is_incremental() %} -- and it has not been processed yet
-  and derived_tstamp > ( select max(last_processed_event) from {{ this }} )
+  where load_tstamp > ( select max(last_processed_event) from {{ this }} )
```
🔧 It's a small table so it shouldn't matter, but is it worth sticking this in a set block above to avoid the subquery filter, just in case?

rlh1994 commented on macros/cluster_by_fields.sql:
```diff
@@ -48,3 +48,15 @@
+
+{% endmacro %}
+
+{% macro default__web_cluster_by_fields_consent() %}
+
+  {{ return(snowplow_utils.get_cluster_by(bigquery_cols=["event_id","domain_userid"], snowflake_cols=["to_date(derived_tstamp)"])) }}
+
+{% endmacro %}
```
Is it worth changing this to `load_tstamp` as you now filter on that more?

agnessnowplow replied:
for sure! I missed that.

rlh1994 reviewed (changes requested):
Just some questions around the cluster key and subquery filters, but looks a lot cleaner now just using the load_tstamp!

rlh1994 reviewed (approved):
LGTM!
